### ISSUE TITLE
compiler rule with golang

### ISSUE THREAD

**Issue by TruElysian** [google/santa#1327]
Hello, I'm attempting to use Santa in lockdown mode with golang.

```
santactl status
>>> Daemon Info
  Mode                      | Lockdown
  Transitive Rules          | Yes
  Log Type                  | file
  File Logging              | Yes
  USB Blocking              | No
  On Start USB Options      | None
  Watchdog CPU Events       | 0  (Peak: 13.88%)
  Watchdog RAM Events       | 1  (Peak: 318.84MB)
>>> Cache Info
  Root cache count          | 55
  Non-root cache count      | 4
>>> Database Info
  Binary Rules              | 30
  Certificate Rules         | 25
  TeamID Rules              | 0
  SigningID Rules           | 23
  CDHash Rules              | 0
  Compiler Rules            | 14
  Transitive Rules          | 8
  Events Pending Upload     | 61
>>> Watch Items
  Enabled                   | No
```

macOS version: `14.4.1`

go version: `go version go1.21.7 darwin/arm64`

santa version:

```
santactl version
santad          | 2024.4 (build 622252801)
santactl        | 2024.4 (build 622252801)
SantaGUI        | 2024.4 (build 622252801)
```

I tried running a go file and here is the output from santa.

```
go run main.go
signal: killed
                                                                                                                                                              
Santa

This application has been blocked from executing.

Path:       /private/var/folders/13/_wt62zh167dd_m1mdzt6lxzw0000gn/T/go-build1412731235/b001/exe/main
Identifier: 929ed1d6baee11bc80d1868f55b1f9517e9e239e86c210a0f8c3a054d03bf1d8
Parent:     go (2292)
```

log

```
action=EXEC|decision=DENY|reason=UNKNOWN|sha256=929ed1d6baee11bc80d1868f55b1f9517e9e239e86c210a0f8c3a054d03bf1d8|pid=2319|pidversion=5499|ppid=2292|uid=501|gid=20|group=staff|mode=L|path=/private/var/folders/13/_wt62zh167dd_m1mdzt6lxzw0000gn/T/go-build1412731235/b001/exe/main|args=/var/folders/13/_wt62zh167dd_m1mdzt6lxzw0000gn/T/go-build1412731235/b001/exe/main
action=DELETE|path=/private/var/folders/13/_wt62zh167dd_m1mdzt6lxzw0000gn/T/go-build1412731235|pid=2292|ppid=2108|process=go|processpath=/opt/homebrew/Cellar/go@1.21/1.21.7/libexec/bin/go|uid=501|gid=20|group=staff
action=DELETE|path=/private/var/folders/13/_wt62zh167dd_m1mdzt6lxzw0000gn/T/go-build1412731235/b001/importcfg.link|pid=2292|ppid=2108|process=go|processpath=/opt/homebrew/Cellar/go@1.21/1.21.7/libexec/bin/go|uid=501|gid=20|group=staff
action=DELETE|path=/private/var/folders/13/_wt62zh167dd_m1mdzt6lxzw0000gn/T/go-build1412731235/b001/exe/main|pid=2292|ppid=2108|process=go|processpath=/opt/homebrew/Cellar/go@1.21/1.21.7/libexec/bin/go|uid=501|gid=20|group=staff
action=DELETE|path=/private/var/folders/13/_wt62zh167dd_m1mdzt6lxzw0000gn/T/go-build1412731235/b001/exe|pid=2292|ppid=2108|process=go|processpath=/opt/homebrew/Cellar/go@1.21/1.21.7/libexec/bin/go|uid=501|gid=20|group=staff
action=DELETE|path=/private/var/folders/13/_wt62zh167dd_m1mdzt6lxzw0000gn/T/go-build1412731235/b001|pid=2292|ppid=2108|process=go|processpath=/opt/homebrew/Cellar/go@1.21/1.21.7/libexec/bin/go|uid=501|gid=20|group=staff
action=DELETE|path=/private/var/folders/13/_wt62zh167dd_m1mdzt6lxzw0000gn/T/go-build1412731235|pid=2292|ppid=2108|process=go|processpath=/opt/homebrew/Cellar/go@1.21/1.21.7/libexec/bin/go|uid=501|gid=20|group=staff
```

I have executed the following command

```shell
# allow rule
sudo santactl rule --allow --path /opt/homebrew/Cellar/go@1.21/1.21.7/libexec/pkg/tool/darwin_arm64/compile
sudo santactl rule --allow --path /opt/homebrew/Cellar/go@1.21/1.21.7/libexec/pkg/tool/darwin_arm64/asm
sudo santactl rule --allow --path /opt/homebrew/Cellar/go@1.21/1.21.7/libexec/pkg/tool/darwin_arm64/cgo
sudo santactl rule --allow --path /opt/homebrew/Cellar/go@1.21/1.21.7/libexec/pkg/tool/darwin_arm64/link

# compiler rule
sudo santactl rule --compiler --identifier EQHXZ8M8AV:org.golang.compile --signingID
sudo santactl rule --compiler --identifier EQHXZ8M8AV:org.golang.link --signingID
sudo santactl rule --compiler --identifier EQHXZ8M8AV:org.golang.asm --signingID
sudo santactl rule --compiler --identifier EQHXZ8M8AV:org.golang.go --signingID
sudo santactl rule --compiler --path ~/DevelopmentTools/go/bin/dlv
sudo santactl rule --compiler --path ~/.vscode-oss/extensions/golang.go-0.41.2/bin/vscgo
sudo santactl rule --compiler --path /opt/homebrew/Cellar/go@1.21/1.21.7/libexec/pkg/tool/darwin_arm64/compile
sudo santactl rule --compiler --path /opt/homebrew/Cellar/go@1.21/1.21.7/libexec/pkg/tool/darwin_arm64/asm
sudo santactl rule --compiler --path /opt/homebrew/Cellar/go@1.21/1.21.7/libexec/pkg/tool/darwin_arm64/cgo
sudo santactl rule --compiler --path /opt/homebrew/Cellar/go@1.21/1.21.7/libexec/pkg/tool/darwin_arm64/link
```

I have read #561 but I don't know how to do it, can you help me please?

Finally, here is my mobileconfig file:

I've modified it from the sample configuration. [com.google.santa.example.mobileconfig](https://santa.dev/deployment/com.google.santa.example.mobileconfig)

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
    <dict>
        <key>PayloadContent</key>
        <array>
            <dict>
                <key>PayloadContent</key>
                <dict>
                    <key>com.google.santa</key>
                    <dict>
                        <key>Forced</key>
                        <array>
                            <dict>
                                <key>mcx_preference_settings</key>
                                <dict>
                                    <key>BannedBlockMessage</key>
                                    <string>This application has been banned</string>
                                    <key>ClientMode</key>
                                    <integer>2</integer>
                                    <key>EnableTransitiveRules</key>
                                    <true />
                                    <key>EnablePageZeroProtection</key>
                                    <false />
                                    <key>EnableSilentMode</key>
                                    <false />
                                    <key>FileChangesRegex</key>
                                    <string>^/(?!(?:private/tmp|Library/(?:Caches|Managed
                                        Installs/Logs|(?:Managed )?Preferences))/)</string>
                                    <key>MachineIDKey</key>
                                    <string>MachineUUID</string>
                                    <key>MachineIDPlist</key>
                                    <string>/Library/Preferences/com.company.machine-mapping.plist</string>
                                    <key>MachineOwnerKey</key>
                                    <string>Owner</string>
                                    <key>MachineOwnerPlist</key>
                                    <string>/Library/Preferences/com.company.machine-mapping.plist</string>
                                    <key>ModeNotificationLockdown</key>
                                    <string>Entering Lockdown mode</string>
                                    <key>ModeNotificationMonitor</key>
                                    <string>Entering Monitor mode&lt;br/&gt;Please be careful!</string>
                                    <key>UnknownBlockMessage</key>
                                    <string>This application has been blocked from executing.</string>
                                </dict>
                            </dict>
                        </array>
                    </dict>
                </dict>
                <key>PayloadEnabled</key>
                <true />
                <key>PayloadIdentifier</key>
                <string>0342c558-a101-4a08-a0b9-40cc00039ea5</string>
                <key>PayloadType</key>
                <string>com.apple.ManagedClient.preferences</string>
                <key>PayloadUUID</key>
                <string>0342c558-a101-4a08-a0b9-40cc00039ea5</string>
                <key>PayloadVersion</key>
                <integer>1</integer>
            </dict>
        </array>
        <key>PayloadDescription</key>
        <string>com.google.santa</string>
        <key>PayloadDisplayName</key>
        <string>com.google.santa</string>
        <key>PayloadIdentifier</key>
        <string>com.google.santa</string>
        <key>PayloadOrganization</key>
        <string></string>
        <key>PayloadRemovalDisallowed</key>
        <true />
        <key>PayloadScope</key>
        <string>System</string>
        <key>PayloadType</key>
        <string>Configuration</string>
        <key>PayloadUUID</key>
        <string>9020fb2d-cab3-420f-9268-acca4868bdd0</string>
        <key>PayloadVersion</key>
        <integer>1</integer>
    </dict>
</plist>
```

**Comment by TruElysian** [comment]
```shell
go build main.go

./main
```

The above commands work fine, but not execute `dlv dap --listen=127.0.0.1:53789 --log-dest=3 from project` in VSCode or execute the `go run main.go` command

**Comment by mlw** [comment]
Transitive rules are tricky to get right, and often involves understanding more about how the specific toolchains in use are generating files to ensure appropriate rules can be created.

One thing that has changed since transitive allowlisting was first introduced is the fact that most every binary is now signed, and on Apple Silicon based macs this is a requirement. We can use this to help make configuring compiler rules a bit easier.

Since code signing is the last operation to be performed on a binary, we can instead create compiler rules targeted at binaries responsible for code signing. On macOS, this is most often `ld(1)` and `codesign(1)`. Can you try creating compiler rules targeting these binaries instead (`com.apple.ld` and `com.apple.security.codesign`)?

**Comment by TruElysian** [comment]
Thanks for your answer, I just executed the following command:

```shell
sudo santactl rule --compiler --path /usr/bin/ld
sudo santactl rule --compiler --path /usr/bin/codesign

sudo santactl rule --compiler --path /Applications/Xcode.app/Contents/MacOS/Xcode
sudo santactl rule --compiler --path /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld
```

```shell
➜ santactl fileinfo /usr/bin/codesign --key "Rule"
Allowed (Compiler)

➜ santactl fileinfo /usr/bin/ld --key "Rule"
Allowed (Compiler)
```

I'm not sure if the `com.apple.ld` you're referring to is `/usr/bin/ld`, I added `/usr/bin/ld` to the `Compiler` and it didn't solve the problem.

```shell
➜ go run main.go
signal: killed

Santa

This application has been blocked from executing.

Path:       /private/var/folders/13/_wt62zh167dd_m1mdzt6lxzw0000gn/T/go-build649072316/b001/exe/main
Identifier: 929ed1d6baee11bc80d1868f55b1f9517e9e239e86c210a0f8c3a054d03bf1d8
Parent:     go (26488)
```

In the meantime, the following command does not produce any output:

```
/usr/bin/log --predicate 'sender=="com.google.santa.daemon"
```

Thanks again for your help.

**Comment by mlw** [comment]
I'm not very familiar with the Go toolchain, but I've spent some time playing around with it and using `eslogger(1)` to track activities such as file creations, modifications, and when the resultant binary is executed (such as under `go run`).

After chatting with some other folks on the project, I think we may understand a bit better about what's going on.

My previous comment seems to not apply very well to the Go toolchain. They appear to not use standard tooling and instead decided to reimplement signing themselves for some unknown reason (https://pkg.go.dev/cmd/internal/codesign). This appears to track with `eslogger(1)` output. This unfortunately means that we cannot use the code signing utilities as stand-ins for compiler rules.

What I observed was that the original compiler rule set you had is largely correct  - though I don't believe the rules for the `compile` and `asm` tools do much. Additionally, the rule for `link` is unfortunately not working as we'd like - more on this later.

In my tests, I made a simple "hello world" go file named `asdfasdf.go` for easy greping. In one terminal, ran the following (with various event sets for whatever was interesting for a given run):
```
sudo eslogger exec mmap open close rename create | grep "asdfasdf"
```
In the other terminal, I ran various Go tools such as:
```
go run asdfasdf.go
go build asdfasdf.go
```

In both the `go build` and `go run` cases, the Go toolchain creates a temporary file, on which various tools are run (`compile`, `asm`, and `link`) to construct the final, signed (at least on Apple Silicon macs) binary.

In the `go build` case, this resultant binary is then moved to the current working directory from which the original command was executed.

In the `go run` case, the resultant temporary binary is directly executed.

When a process executes that matches a compiler rule, Santa will monitor various operations of that process in order to create transitive rules. One such operation is `rename(2)`. When tracing with `eslogger(1)` while performing a `go build`, one of the last steps performed as mentioned above is moving the temporary file in its final form to the current directory. This is done by the main Go binary (`org.golang.org`). When this happens, Santa creates a binary rule for that file. Since that file is already signed and will no longer be modified, it's hash is correct and will not change. When you execute this file, the transitive rule is already in place and things work as expected.

For `go run` however, there is a problem. Signing operations are performed by the `link` tool (`org.golang.link`). The set of operations that Santa monitors for tracking compiler operations are all asynchronous. Because `go run` immediately executes the binary, it appears that there is a race condition between when Santa processes the event (in this case "close-modified" or "close-was-mapped-wrtiable") which will generate the rule, and the binary being executed. When the `EXEC` event happens, this is authorized on a separate event stream within Santa. If the transitive rule hasn't yet been created (e.g. because the event stream tracking compiler operations hasn't processed that event yet), then the execution will be denied.

Further, the Go toolchain appears to perform file modifications on mapped files, complicating things a bit. Santa currently doesn't track `munmap(2)` operations. However, Go toolchain appears to do things in a slightly non-idiomatic way and doesn't close the file immediately after `mmap(2)`ing. This at least allows Santa to see the file in the final form after a `close(2)` operation. However, we should still likely begin tracking `munmap(2)` as well.

What this means: The transitive allowlisting feature does not currently support toolchains that feature operations like `go run` (or `cargo run`) - things that create and immediately execute these executables.

The workaround is to produce the intermediate binary (e.g. via `go build`) and then to execute that. We'll need to think about a longer term approach here to better support toolchains that have these sorts of operations.
